clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java

```diff
@@ -175,6 +173,60 @@ public void testBlockTimeout() throws Exception {
     }
 
     /**
+     * Test if the  waiter that is waiting on availability of more memory is cleaned up when a timeout occurs
+     */
+    @Test
+    public void testCleanupMemoryAvailabilityWaiterOnBlockTimeout() throws Exception {
+        BufferPool pool = new BufferPool(2, 1, metrics, time, metricGroup);
+        long blockTime = 5000;
+        pool.allocate(1, maxBlockTimeMs);
+        Thread t1 = new Thread(new BufferPoolAllocator(pool, blockTime));
+        Thread t2 = new Thread(new BufferPoolAllocator(pool, blockTime));
+        // allocate
+        t1.start();
+        // sleep for 1 sec, the waiters queue will be populated with the condition variable for the allocate by thread t1
+        Thread.sleep(1000);
+        Deque<Condition> waiters =  pool.waiters();
+        // get the condition object associated with allocate() by thread t1
+        Condition c1 = waiters.getFirst();
+        // sleep for 4 sec, the waiter queue should have been cleared by now since the blockTime threshold must have kicked in
+        Thread.sleep(4000);
```

Comment:
Could we use a lower sleep and block time to make the test faster? Or would it make it unreliable?